### PR TITLE
Implement __eq__ for DL3 IRFs

### PR DESCRIPTION
**Pull request by AtreyeeS**
Addresses  #3534

### CODE REVIEW

**Review by maxnoe — COMMENTED** [review]
Why the indirection using a `compare` method? Why not just directly implement `__eq__`?

**Comment by adonath** [comment]
@maxnoe I think the main argument is that a `.compare()` method can take arguments such as relative and absolute tolerance, which is probably required, depending on the use case. However we should maybe think a bit about a uniform pattern. Similar methods are here:
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/axes.py#L177
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/axes.py#L153
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/axes.py#L2276
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/axes.py#L2912
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/wcs/geom.py#L1129
- https://github.com/gammapy/gammapy/blob/master/gammapy/maps/wcs/geom.py#L1107

Maybe a method like `IRF.is_allclose(other, atol, rtol)`  and `IRF.axes.is_allclose(other, atol, rtol)` is more flexible for testing. `__eq__` would then be implemented as `self.is_allcose(other, atol=1e-12, rtol=1e-6, atol_axes=, rtol_axes=), with some default tolerance.

File: gammapy/irf/core.py [hunk 1]
```diff
@@ -546,6 +554,31 @@ def pad(self, pad_width, axis_name, **kwargs):
...
+
+        kwargs.setdefault("rtol", 1e-3)
+        if self.data.shape != other.data.shape:
+            return False
+        return np.allclose(self.quantity, other.quantity, **kwargs)
+
+    def __eq__(self, other):
+        class_eq = isinstance(other, self.__class__)
+        axes_eq = self.axes == other.axes
+        return bool(class_eq * axes_eq * self.data_allclose(other))
```
Comment (registerrier) [inline gammapy/irf/core.py:580]: why not using `and` ?

File: gammapy/irf/core.py [hunk 2]
```diff
@@ -546,6 +554,31 @@ def pad(self, pad_width, axis_name, **kwargs):
...
 
+    def data_allclose(self, other, **kwargs):
+        """Compare two data IRFs for equivalency
+
+        Parameters
+        ----------
+        other: the irf to compare against
+            `gammapy.irfs.IRF`
+        kwargs: dict
+                keywords passed to `numpy.allclose`
```
Comment (registerrier) [inline gammapy/irf/core.py:565]: maybe specify the default `rtol`?

File: gammapy/irf/core.py [hunk 3]
```diff
@@ -546,6 +554,31 @@ def pad(self, pad_width, axis_name, **kwargs):
...
+
+        Returns
+        -------
+        state: bool
```
Comment (registerrier) [inline gammapy/irf/core.py:569]: space missing: `state :  bool`

File: gammapy/irf/psf/parametric.py [hunk 4]
```diff
@@ -248,6 +248,29 @@ def evaluate(self, rad, **kwargs):
...
+        Parameters
+        ----------
+        other: the irf to compare against
+                    `gammapy.irfs.IRF`
+        kwargs: dict
+                keywords passed to `numpy.allclose`
+
+        Returns
+        -------
+        state: bool
```
Comment (registerrier) [inline gammapy/irf/psf/parametric.py:263]: space missing.

File: gammapy/irf/psf/parametric.py [hunk 5]
```diff
@@ -248,6 +248,29 @@ def evaluate(self, rad, **kwargs):
...
+        """
+
+        state = True
+        for key in self.quantity.keys():
+            if self.quantity[key].shape != other.quantity[key].shape:
```
Comment (registerrier) [inline gammapy/irf/psf/parametric.py:268]: Here you might simply return False

File: gammapy/irf/psf/parametric.py [hunk 6]
```diff
@@ -248,6 +248,29 @@ def evaluate(self, rad, **kwargs):
...
+                state = False
+            state *= np.allclose(self.quantity[key], other.quantity[key], **kwargs)
```
Comment (registerrier) [inline gammapy/irf/psf/parametric.py:270]: Here also you can only return False as soon as the `allclose` is False.

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS . This looks good.
I have left some small suggestions.

File: gammapy/irf/core.py [hunk 7]
```diff
@@ -546,6 +554,32 @@ def pad(self, pad_width, axis_name, **kwargs):
...
+
+        kwargs.setdefault("rtol", 1e-3)
+        if self.data.shape != other.data.shape:
+            return False
+        return np.allclose(self.quantity, other.quantity, **kwargs)
+
+    def __eq__(self, other):
+        class_eq = isinstance(other, self.__class__)
+        axes_eq = self.axes == other.axes
+        return bool(class_eq & axes_eq & self.data_allclose(other))
```
Comment (registerrier) [inline gammapy/irf/core.py:581]: `&` is a bitwise operator. Here it should work but I think the correct syntax should be:
`class_eq and axes_eq and self.data_allclose(other)`

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS

**Pull request merged**